File: src/opnsense/scripts/filter/lib/alias.py
"""
    Copyright (c) 2017-2021 Ad Schellevis <ad@opnsense.org>
    All rights reserved.

    Redistribution and use in source and binary forms, with or without
    modification, are permitted provided that the following conditions are met:

    1. Redistributions of source code must retain the above copyright notice,
     this list of conditions and the following disclaimer.

    2. Redistributions in binary form must reproduce the above copyright
     notice, this list of conditions and the following disclaimer in the
     documentation and/or other materials provided with the distribution.

    THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES,
    INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
    AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY,
    OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
    SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
    INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
    CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
    ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
    POSSIBILITY OF SUCH DAMAGE.

    --------------------------------------------------------------------------------------
    Alias representation
"""
import os
import re
import time
import requests
import ipaddress
import dns.resolver
import syslog
from hashlib import md5
from . import geoip
from . import net_wildcard_iterator
from .arpcache import ArpCache

class Alias(object):
    def __init__(self, elem, known_aliases=[], ttl=-1, ssl_no_verify=False, timeout=120):
        """ construct alias object
            :param elem: ElementTree alias item
            :param known_aliases: all known alias names
            :param ttl: time to live in seconds (for other then ip/network types)
            :param ssl_no_verify: disable ssl verify when fetching content
            :param timeout: request timeout in seconds
            :return: None
        """
        self._known_aliases = known_aliases
        self._dnsResolver = dns.resolver.Resolver()
        self._dnsResolver.timeout = 2
        self._is_changed = None
        self._has_expired = None
        self._ttl = ttl
        self._ssl_no_verify = ssl_no_verify
        self._timeout = timeout
        self._name = None
        self._type = None
        self._proto = 'IPv4,IPv6'
        self._items = list()
        self._resolve_content = set()
        for subelem in elem:
            if subelem.tag == 'type':
                self._type = subelem.text
            elif subelem.tag == 'proto':
                self._proto = subelem.text
            elif subelem.tag == 'name':
                self._name = subelem.text
            elif subelem.tag == 'ttl':
                tmp = subelem.text.strip()
                if len(tmp.split('.')) <= 2 and tmp.replace('.', '').isdigit():
                    self._ttl = int(float(tmp))
            elif subelem.tag in ('aliasurl', 'address', 'url') and subelem.text is None:
                self._items = set()
            elif subelem.tag == 'aliasurl':
                self._items = set(sorted(subelem.text.split()))
            elif subelem.tag == 'address' and len(self._items) == 0:
                # special case, aliasurl fetched addresses in old implentation we don't want to use them
                self._items = set(sorted(subelem.text.split()))
            elif subelem.tag == 'url':
                self._items = set(sorted(subelem.text.split()))
        # we'll save the calculated hash for the unparsed alias content
        self._filename_alias_hash = '/var/db/aliastables/%s.md5.txt' % self._name
        # the generated alias contents, without dependencies
        self._filename_alias_content = '/var/db/aliastables/%s.self.txt' % self._name

    def _parse_address(self, address):
        """ parse addresses and hostnames, yield only valid addresses and networks
            :param address: address or network
            :return: boolean
        """
        address = address.strip()
        if address.find('/') > -1 and not address.split('/')[-1].isdigit():
            # wildcard netmask
            for idx, item in enumerate(net_wildcard_iterator(address.lstrip('!'))):
                if idx > 65535:
                    # overflow
                    syslog.syslog(syslog.LOG_ERR, 'alias table %s overflow' % self._name)
                    break
                yield "!%s" % item if address.startswith('!') else str(item)
        elif address.find('/') > -1:
            # provided address could be a network
            try:
                ipaddress.ip_network(str(address.lstrip('!')), strict=False)
                yield address
                return
            except (ipaddress.AddressValueError, ValueError):
                pass
        else:
            # check if address is an ipv4/6 address or range
            try:
                tmp = str(address).split('-')
                if len(tmp) > 1:
                    addr1 = ipaddress.ip_address(tmp[0])
                    # address range (from-to)
                    addr2 = ipaddress.ip_address(tmp[1])
                    for addr in ipaddress.summarize_address_range(addr1, addr2):
                        yield str(addr)
                else:
                    ipaddress.ip_address(tmp[0].lstrip('!'))
                    yield address
                return
            except (ipaddress.AddressValueError, ValueError):
                pass

        # try to resolve provided address
        could_resolve = False
        for record_type in ['A', 'AAAA']:
            try:
                for rdata in self._dnsResolver.query(address, record_type):
                    yield str(rdata)
                could_resolve = True
            except (dns.resolver.NoAnswer, dns.resolver.NXDOMAIN, dns.exception.Timeout, dns.resolver.NoNameservers, dns.name.EmptyLabel):
                pass

        if not could_resolve:
            # log when none could be found
            syslog.syslog(syslog.LOG_ERR, 'unable to resolve %s for alias %s' % (address, self._name))

    def _fetch_url(self, url):
        """ return unparsed (raw) alias entries without dependencies
            :param url: url
            :return: iterator
        """
        # set request parameters
        req_opts = dict()
        req_opts['url'] = url
        req_opts['stream'] = True
        req_opts['timeout'] = self._timeout
        if self._ssl_no_verify:
            req_opts['verify'] = False
        # fetch data
        try:
            req = requests.get(**req_opts)
            if req.status_code == 200:
                # only handle content if response is correct
                req.raw.decode_content = True
                lines = req.raw.read().decode().splitlines()
                syslog.syslog(syslog.LOG_NOTICE, 'fetch alias url %s (lines: %s)' % (url, len(lines)))
                for line in lines:
                    raw_address = re.split(r'[\s,;|#]+', line)[0]
                    if raw_address and not raw_address.startswith('//'):
                        for address in self._parse_address(raw_address):
                            yield address
            else:
                syslog.syslog(syslog.LOG_ERR, 'error fetching alias url %s [http_code:%s]' % (url, req.status_code))
                raise IOError('error fetching alias url %s' % (url))
        except:
            syslog.syslog(syslog.LOG_ERR, 'error fetching alias url %s' % (url))
            raise IOError('error fetching alias url %s' % (url))

    def _fetch_geo(self, geoitem):
        """ fetch geoip addresses, if not downloaded or outdated force an update
            :return: iterator
        """
        do_update = True
        if os.path.isfile('/usr/local/share/GeoIP/alias/NL-IPv4'):
            fstat = os.stat('/usr/local/share/GeoIP/alias/NL-IPv4')
            if (time.time() - fstat.st_mtime) < (86400 - 90):
                do_update = False
        if do_update:
            syslog.syslog(syslog.LOG_ERR, 'geoip updated (files: %(file_count)d lines: %(address_count)d)' % geoip.download_geolite())

        for proto in self._proto.split(','):
            geoip_filename = "/usr/local/share/GeoIP/alias/%s-%s" % (geoitem, proto)
            if os.path.isfile(geoip_filename):
                with open(geoip_filename) as f_in:
                    for line in f_in:
                        for address in self._parse_address(line):
                            yield address

    def items(self):
        """ return unparsed (raw) alias entries without dependencies
            :return: iterator
        """
        for item in self._items:
            if item not in self._known_aliases or self.get_name() == item:
                yield item

    def uniqueid(self):
        """ generate an identification hash for this alias
            :return: md5 (string)
        """
        tmp = ','.join(sorted(list(self._items)))
        if self._proto:
            tmp = '%s[%s]' % (tmp, self._proto)
        return md5(tmp.encode()).hexdigest()

    def changed(self):
        """ is the alias changed (cached result, if changed within this objects lifetime)
            :return: boolean
        """
        if self._is_changed is None:
            if os.path.isfile(self._filename_alias_hash) and os.path.isfile(self._filename_alias_content):
                self._is_changed = open(self._filename_alias_hash).read().strip() != self.uniqueid()
            else:
                self._is_changed = True

        return self._is_changed

    def expired(self):
        """ if this alias has an expiry (ttl), has it reached the end of it's lifetime
            :return: boolean
        """
        if self._has_expired is None:
            if self._ttl > 0 and os.path.isfile(self._filename_alias_hash):
                fstat = os.stat(self._filename_alias_hash)
                self._has_expired = time.time() - fstat.st_mtime > self._ttl
            else:
                self._has_expired = False
        return self._has_expired

    def resolve(self, force=False):
        """ resolve (fetch) alias content, without dependencies.
            :force: force load
            :return: string
        """
        if not self._resolve_content:
            if self.expired() or self.changed() or force:
                if os.path.isfile(self._filename_alias_content):
                    undo_content = open(self._filename_alias_content, 'r').read()
                else:
                    undo_content = ""
                try:
                    with open(self._filename_alias_content, 'w') as f_out:
                        for item in self.items():
                            address_parser = self.get_parser()
                            if address_parser:
                                for address in address_parser(item):
                                    if address not in self._resolve_content:
                                        # flush new alias content (without dependencies) to disk, so progress can easliy
                                        # be followed, large lists of domain names can take quite some resolve time.
                                        f_out.write('%s\n' % address)
                                        f_out.flush()
                                        # preserve addresses
                                        self._resolve_content.add(address)
                except IOError:
                    # parse issue, keep data as-is, flush previous content to disk
                    with open(self._filename_alias_content, 'w') as f_out:
                        f_out.write(undo_content)
                    self._resolve_content = set(undo_content.split("\n"))
                # flush md5 hash to disk
                open(self._filename_alias_hash, 'w').write(self.uniqueid())
            else:
                self._resolve_content = set(open(self._filename_alias_content).read().split())
        # return the addresses and networks of this alias
        return list(self._resolve_content)

    def get_parser(self):
        """ fetch address parser to use, None if alias type is not handled here
            :return: function or None
        """
        if self._type in ['host', 'network', 'networkgroup']:
            return self._parse_address
        elif self._type in ['url', 'urltable']:
            return self._fetch_url
        elif self._type == 'geoip':
            return self._fetch_geo
        elif self._type == 'mac':
            return ArpCache().iter_addresses
        else:
            return None

    def get_type(self):
        """ get type of alias
            :return: string
        """
        return self._type

    def get_name(self):
        """ get alias name
            :return: string
        """
        return self._name

    def get_deps(self):
        """ fetch alias dependencies
            :param in_data: raw input data (ruleset)
            :return: new ruleset
        """
        for item in self._items:
            if item in self._known_aliases:
                yield item

File: src/opnsense/mvc/app/models/OPNsense/Firewall/Alias.php
<?php

/*
 * Copyright (C) 2018 Deciso B.V.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice,
 *    this list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES,
 * INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY
 * AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
 * AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY,
 * OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

namespace OPNsense\Firewall;

use OPNsense\Base\BaseModel;
use OPNsense\Core\Config;
use OPNsense\Firewall\Util;
use OPNsense\Base\Messages\Message;

/**
 * Class Alias
 * @package OPNsense\Firewall
 */
class Alias extends BaseModel
{
    /**
     * cache iteration items
     */
    private $aliasIteratorCache = [];

    /**
     * alias name cache
     */
    private $aliasReferenceCache = [];

    public function performValidation($validateFullModel = false)
    {
        $messages = parent::performValidation($validateFullModel);

        foreach ($this->aliases->alias->iterateItems() as $alias) {
            if (!$validateFullModel && !$alias->isFieldChanged()) {
                continue;
            }

            $ref = $alias->__reference;

            $username = (string)$alias->username;
            $password = (string)$alias->password;

            switch ((string)$alias->authtype) {
                case 'Basic':
                    if (empty($username) || empty($password)) {
                        $messages->appendMessage(new Message(gettext('Please provide a username and password when Basic auth is selected'), $ref . '.authtype'));
                    }
                    break;
                case 'Bearer':
                    if (empty($password)) {
                        $messages->appendMessage(new Message(gettext('Please provide an API token when Bearer auth is selected'), $ref . '.authtype'));
                    } elseif (strlen($password) > 512) {
                        $messages->appendMessage(new Message(gettext('Invalid token length'), $ref . '.authtype'));
                    } elseif (!preg_match('/^[A-Za-z0-9-_.]+$/', $password)) {
                        $messages->appendMessage(new Message(gettext('Illegal characters in token'), $ref . '.authtype'));
                    }
                    break;
            }
        }

        return $messages;
    }

    /**
     * return locations where aliases can be used inside the configuration
     * @return array alias source map
     */
    private function getAliasSource()
    {
        $sources = [];
        $sources[] = [['filter', 'rule'], ['source', 'address']];
        $sources[] = [['filter', 'rule'], ['destination', 'address']];
        $sources[] = [['filter', 'rule'], ['source', 'port']];
        $sources[] = [['filter', 'rule'], ['destination', 'port']];
        $sources[] = [['filter', 'scrub', 'rule'], ['dst']];
        $sources[] = [['filter', 'scrub', 'rule'], ['src']];
        $sources[] = [['filter', 'scrub', 'rule'], ['dstport']];
        $sources[] = [['filter', 'scrub', 'rule'], ['srcport']];
        $sources[] = [['nat', 'rule'], ['source', 'address']];
        $sources[] = [['nat', 'rule'], ['source', 'port']];
        $sources[] = [['nat', 'rule'], ['destination', 'address']];
        $sources[] = [['nat', 'rule'], ['destination', 'port']];
        $sources[] = [['nat', 'rule'], ['target']];
        $sources[] = [['nat', 'rule'], ['local-port']];
        $sources[] = [['nat', 'onetoone'], ['destination', 'address']];
        $sources[] = [['nat', 'outbound', 'rule'], ['source', 'network']];
        $sources[] = [['nat', 'outbound', 'rule'], ['sourceport']];
        $sources[] = [['nat', 'outbound', 'rule'], ['destination', 'network']];
        $sources[] = [['nat', 'outbound', 'rule'], ['dstport']];
        $sources[] = [['nat', 'outbound', 'rule'], ['target']];
        $sources[] = [['staticroutes', 'route'], ['network']];
        // os-firewall plugin paths
        $sources[] = [['OPNsense', 'Firewall', 'Filter', 'rules', 'rule'], ['source_net']];
        $sources[] = [['OPNsense', 'Firewall', 'Filter', 'rules', 'rule'], ['source_port']];
        $sources[] = [['OPNsense', 'Firewall', 'Filter', 'rules', 'rule'], ['destination_net']];
        $sources[] = [['OPNsense', 'Firewall', 'Filter', 'rules', 'rule'], ['destination_port']];
        $sources[] = [['OPNsense', 'Firewall', 'Filter', 'snatrules', 'rule'], ['source_net']];
        $sources[] = [['OPNsense', 'Firewall', 'Filter', 'snatrules', 'rule'], ['source_port']];
        $sources[] = [['OPNsense', 'Firewall', 'Filter', 'snatrules', 'rule'], ['destination_net']];
        $sources[] = [['OPNsense', 'Firewall', 'Filter', 'snatrules', 'rule'], ['destination_port']];

        return $sources;
    }

    /**
     * search configuration items where the requested alias is used
     * @param $name alias name
     * @return \Generator [reference, confignode, matched item]
     */
    private function searchConfig($name)
    {
        $cfgObj = Config::getInstance()->object();
        foreach ($this->getAliasSource() as $aliasref) {
            $cfgsection = $cfgObj;
            foreach ($aliasref[0] as $cfgName) {
                if ($cfgsection != null) {
                    $cfgsection = $cfgsection->$cfgName;
                }
            }
            if ($cfgsection != null) {
                $nodeidx = 0;
                foreach ($cfgsection as $inode) {
                    $node = $inode;
                    foreach ($aliasref[1] as $cfgName) {
                        $node = $node->$cfgName;
                    }
                    if ((string)$node == $name) {
                        $ref = implode('.', $aliasref[0]) . "." . $nodeidx . "/" . implode('.', $aliasref[1]);
                        yield array($ref, &$inode, &$node);
                    }
                    $nodeidx++;
                }
            }
        }
    }

    /**
     * flush cached objects and references
     */
    public function flushCache()
    {
        $this->aliasIteratorCache = [];
        $this->aliasReferenceCache = [];
    }

    /**
     * Return all places an alias seems to be used
     * @param string $name alias name
     * @return array hashmap with references where this alias is used
     */
    public function whereUsed($name)
    {
        $result = array();
        // search legacy locations
        foreach ($this->searchConfig($name) as $item) {
            $result[$item[0]] = (string)$item[1]->descr;
        }
        // find all used in this model
        foreach ($this->aliases->alias->iterateItems() as $alias) {
            if (!in_array($alias->type, array('geoip', 'urltable'))) {
                $nodeData = $alias->content->getNodeData();
                if (isset($nodeData[$name])) {
                    if (!empty($alias->description)) {
                        $result[(string)$alias->__reference] = (string)$alias->description;
                    } else {
                        $result[(string)$alias->__reference] = (string)$alias->name;
                    }
                }
            }
        }
        return $result;
    }

    /**
     * replace alias usage
     * @param $oldname
     * @param $newname
     */
    public function refactor($oldname, $newname)
    {
        Util::attachAliasObject($this);
        // replace in legacy config
        foreach ($this->searchConfig($oldname) as $item) {
            $item[2][0] = $newname;
        }
        // find all used in this model (alias nesting)
        foreach ($this->aliases->alias->iterateItems() as $alias) {
            if (!in_array($alias->type, array('geoip', 'urltable'))) {
                $sepchar = $alias->content->getSeparatorChar();
                $aliases = explode($sepchar, (string)$alias->content);
                if (in_array($oldname, $aliases)) {
                    $aliases = array_unique($aliases);
                    $aliases[array_search($oldname, $aliases)] = $newname;
                    $alias->content->setValue(implode($sepchar, $aliases));
                }
            }
        }
    }

    /**
     * return aliases as array
     * @param $flush flush cached objects from previous call
     * @return Generator with aliases
     */
    public function aliasIterator($flush = false)
    {
        if ($flush) {
            // flush cache
            $this->aliasIteratorCache = [];
        }
        foreach ($this->aliases->alias->iterateItems() as $uuid => $alias) {
            if (empty($this->aliasIteratorCache[$uuid])) {
                $this->aliasIteratorCache[$uuid] = [];
                foreach ($alias->iterateItems() as $key => $value) {
                    $this->aliasIteratorCache[$uuid][$key] = (string)$value;
                }
                // parse content into separate items for easier reading. items are separated by \n
                if ((string)$alias->content == "") {
                    $this->aliasIteratorCache[$uuid]['content'] = [];
                } else {
                    $this->aliasIteratorCache[$uuid]['content'] = explode("\n", (string)$alias->content);
                }
            }
            yield $this->aliasIteratorCache[$uuid];
        }
    }

    public function getByName($name, $flush = false)
    {
        if ($flush) {
            // Flush cache in case model data has been changed.
            $this->aliasReferenceCache = [];
        }
        if (empty($this->aliasReferenceCache)) {
            // cache alias uuid references, but always validate existence before return.
            foreach ($this->aliases->alias->iterateItems() as $uuid => $alias) {
                $this->aliasReferenceCache[(string)$alias->name] = $alias;
            }
        }
        if (isset($this->aliasReferenceCache[$name])) {
            return $this->aliasReferenceCache[$name];
        } else {
            return null;
        }
    }

    /**
     * Resolves an alias name (recursively) into a list of network entities (IPs/CIDRs).
     * Filters out non-network types. Enforces recursion depth and entry limits.
     *
     * @param string $name The alias name to resolve.
     * @param int $limit Maximum number of entries to return. Default 1000.
     * @param array $aliases Aliases already parsed (recursion guard).
     * @param int $recursion_depth Current depth (recursion guard). Max depth e.g., 10.
     * @return array|false List of unique IP/CIDR strings on success, false on error/limit/recursion.
     */
    public static function resolveAliasToNetworks($name, $limit = 1000, $aliases = [], $recursion_depth = 0)
    {
        // --- Guards ---
        if ($recursion_depth > 10) {
            syslog(LOG_WARNING, 'OPNsense\Firewall\Alias::resolveAliasToNetworks: Max recursion depth reached resolving \'' . $name . '\'.');
            return false;
        }
        if (in_array($name, $aliases)) {
            syslog(LOG_WARNING, 'OPNsense\Firewall\Alias::resolveAliasToNetworks: Circular reference detected for alias \'' . $name . '\'.');
            return false; // Prevent infinite loop
        }
        $aliases[] = $name;

        // --- Get Alias Node ---
        try {
            // Create an instance to use non-static methods
            $aliasModel = new self();
            // Flush the cache once at the beginning of the resolver
            if ($recursion_depth == 0) {
                $aliasModel->flushCache();
            }
            $aliasNode = $aliasModel->getByName($name);
        } catch (\Exception $e) {
            syslog(LOG_ERR, 'OPNsense\Firewall\Alias::resolveAliasToNetworks: Exception fetching alias \'' . $name . '\': ' . $e->getMessage());
            return false;
        }

        if ($aliasNode == null) {
            syslog(LOG_WARNING, 'OPNsense\Firewall\Alias::resolveAliasToNetworks: Alias \'' . $name . '\' not found.');
            return false;
        }

        // --- Type Check ---
        // Define types that are expected to resolve to IPs/Networks for shaper usage
        $allowed_types = ['host', 'network', 'urltable', 'geoip', 'external'];
        $current_type = (string)$aliasNode->type;
        if (!in_array($current_type, $allowed_types)) {
            // syslog(LOG_DEBUG, "Alias '{$name}' of type '{$current_type}' skipped (not a network type).");
            return []; // Not an error for the caller, just not relevant
        }

        // --- Resolve Content ---
        $resolved_content_map = []; // Use map keys for uniqueness
        try {
            // Get alias content based on type
            $raw_items = [];

            if (in_array($current_type, ['host', 'network'])) {
                // Simple case: get content from node
                if ((string)$aliasNode->content == "") {
                    $raw_items = [];
                } else {
                    $raw_items = explode("\n", (string)$aliasNode->content);
                }
            } elseif ($current_type == 'urltable') {
                // For URL tables, we would need to fetch and parse the content
                // TODO: Implement proper URL table resolution in a future PR
                syslog(LOG_WARNING, 'OPNsense\Firewall\Alias::resolveAliasToNetworks: URLTable resolution for \'' . $name . '\' not fully implemented yet. Skipping.');
                $raw_items = [];
            } elseif ($current_type == 'geoip') {
                // For GeoIP, we would need to lookup the database
                // TODO: Implement proper GeoIP resolution in a future PR
                syslog(LOG_WARNING, 'OPNsense\Firewall\Alias::resolveAliasToNetworks: GeoIP resolution for \'' . $name . '\' not fully implemented yet. Skipping.');
                $raw_items = [];
            } elseif ($current_type == 'external') {
                // External type handling
                // TODO: Implement proper external alias resolution in a future PR
                syslog(LOG_WARNING, 'OPNsense\Firewall\Alias::resolveAliasToNetworks: External resolution for \'' . $name . '\' not fully implemented yet. Skipping.');
                $raw_items = [];
            }

            foreach ($raw_items as $item) {
                $item = trim($item);
                if (empty($item) || strpos($item, '#') === 0) continue; // Skip empty lines and comments

                // Remove potential comments after address (common in URL tables)
                $item = trim(explode('#', $item)[0]);

                if (Util::isIpAddress($item) || Util::isSubnet($item)) {
                    $resolved_content_map[$item] = 1;
                } elseif (Util::isAlias($item)) { // Nested alias
                    // Pass the current limit minus already found items
                    $current_limit = $limit - count($resolved_content_map);
                    if ($current_limit <= 0) {
                        syslog(LOG_WARNING, 'OPNsense\Firewall\Alias::resolveAliasToNetworks: Alias \'' . $name . '\' expansion limit reached before resolving nested alias \'' . $item . '\'.');
                        return false; // Exceeded limit
                    }
                    $nested_result = self::resolveAliasToNetworks($item, $current_limit, $aliases, $recursion_depth + 1);
                    if ($nested_result === false) return false; // Propagate error
                    foreach ($nested_result as $nested_item) {
                        $resolved_content_map[$nested_item] = 1;
                    }
                }
                // Else: Ignore items that are not IPs, Subnets, or resolvable Aliases

                // --- Limit Check ---
                if (count($resolved_content_map) > $limit) {
                    syslog(LOG_WARNING, 'OPNsense\Firewall\Alias::resolveAliasToNetworks: Alias \'' . $name . '\' expansion exceeded limit of ' . $limit . '.');
                    return false;
                }
            }
        } catch (\Exception $e) {
            syslog(LOG_ERR, 'OPNsense\Firewall\Alias::resolveAliasToNetworks: Exception processing content for alias \'' . $name . '\': ' . $e->getMessage());
            return false;
        }

        if (empty($resolved_content_map) && in_array($current_type, ['host', 'network'])) {
            // Log potentially unexpected empty result for static types
            syslog(LOG_NOTICE, 'OPNsense\Firewall\Alias::resolveAliasToNetworks: Alias \'' . $name . '\' of type \'' . $current_type . '\' resolved to an empty list.');
        }

        // --- Return ---
        return array_keys($resolved_content_map);
    }
}
